csharp/src/Ice/Reference.cs

```diff
@@ -745,149 +868,143 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
+                            // preserve any previous exceptions that might have been throw.
+                            Debug.Assert(connectors == null || connectors.Count == 0);
                             childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
-                        }
-                        catch (TransportException ex)
-                        {
-                            var closedException = ex as ConnectionClosedException;
-                            connector ??= ex.Connector;
-                            if (connector != null && closedException == null)
-                            {
-                                reference.Communicator.OutgoingConnectionFactory.AddTransportFailure(connector);
-                            }
-
-                            lastException = ex;
-                            childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
-
-                            // Retry transport exceptions if the request is idempotent, was not sent or if the
-                            // connection was gracefully closed by the peer (in which case it's safe to retry).
-                            if ((closedException?.IsClosedByPeer ?? false) || request.IsIdempotent || !sent)
-                            {
-                                retryPolicy = ex.RetryPolicy;
-                            }
+                            return response ?? throw exception ?? ex;
                         }
                         catch (Exception ex)
                         {
-                            lastException = ex;
+                            exception = ex;
                             childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
                         }
                         finally
                         {
                             childObserver?.Detach();
                         }
 
-                        if (sent && releaseRequestAfterSent)
+                        // Compute retry policy based on the exception or response retry policy, whether or not the connection
+                        // is established or the request sent and idempotent
+                        Debug.Assert(response != null || exception != null);
+                        RetryPolicy retryPolicy =
+                            response?.GetRetryPolicy(reference) ??
+                            exception!.GetRetryPolicy(request.IsIdempotent, sent);
+
+                        // With the retry-policy OtherReplica we add the current connector to the list of excluded
+                        // connectors, this prevents the connector to be tried again during the current retry sequence.
+                        if (retryPolicy == RetryPolicy.OtherReplica)
                         {
-                            if (reference.Communicator.TraceLevels.Retry >= 1)
+                            if (connectors != null)
```

Comment:
It could also be:
```
                    if ((connectors?[nextConnector] ?? connection?.Connector) is Connector connector)
                    {
                        excludedConnectors ??= new ();
                        excludedConnectors.Add(connector);
                    }
                    connectors?.RemoveAt(nextConnector);
```